lib/vnet/ssh_proxy.go

```diff
@@ -0,0 +1,254 @@
+		log.DebugContext(ctx, "Failed to accept SSH channel request already accepted by the target, killing the connection",
+			"error", err)
+		if err := targetChan.Close(); err != nil {
+			log.DebugContext(ctx, "Failed to close SSH channel on target",
+				"error", err)
+		}
+		closeConnections()
+		return
+	}
+
+	// Copy channel requests in both directions concurrently. If either fails or
+	// exits it will cancel the context so that utils.ProxyConn below will close
+	// both channels so the other goroutine can also exit.
+	var wg sync.WaitGroup
+	wg.Add(2)
+	ctx, cancel := context.WithCancel(ctx)
+	go func() {
+		proxyChannelRequests(ctx, log, targetChan, incomingChanRequests, cancel)
+		cancel()
+		wg.Done()
+	}()
+	go func() {
+		proxyChannelRequests(ctx, log, incomingChan, targetChanRequests, cancel)
+		cancel()
+		wg.Done()
+	}()
+
+	// ProxyConn copies channel data bidirectionally. If the context is
+	// canceled it will terminate, it always closes both channels before
+	// returning.
+	if err := utils.ProxyConn(ctx, incomingChan, targetChan); err != nil &&
+		!utils.IsOKNetworkError(err) && !errors.Is(err, context.Canceled) {
+		log.DebugContext(ctx, "Unexpected error proxying channel data", "error", err)
+	}
+
+	// Wait for all goroutines to terminate.
+	wg.Wait()
+}
+
+func proxyChannelRequests(
+	ctx context.Context,
+	log *slog.Logger,
+	targetChan ssh.Channel,
+	reqs <-chan *ssh.Request,
+	closeChannels func(),
+) {
+	log = log.With("request_layer", "channel")
+	sendRequest := func(name string, wantReply bool, payload []byte) (bool, []byte, error) {
+		ok, err := targetChan.SendRequest(name, wantReply, payload)
+		// Replies to channel requests never have a payload.
+		return ok, nil, err
+	}
+	proxyRequests(ctx, log, sendRequest, reqs, closeChannels)
+}
+
+func proxyGlobalRequests(
+	ctx context.Context,
+	targetConn ssh.Conn,
+	reqs <-chan *ssh.Request,
+	closeConnections func(),
+) {
+	log := log.With("request_layer", "global")
+	sendRequest := targetConn.SendRequest
+	proxyRequests(ctx, log, sendRequest, reqs, closeConnections)
+}
+
+func proxyRequests(
+	ctx context.Context,
+	log *slog.Logger,
+	sendRequest func(name string, wantReply bool, payload []byte) (bool, []byte, error),
+	reqs <-chan *ssh.Request,
+	closeRequestSources func(),
+) {
+	for req := range reqs {
+		log := log.With("request_type", req.Type)
+		log.DebugContext(ctx, "Proxying SSH request")
+		ok, reply, err := sendRequest(req.Type, req.WantReply, req.Payload)
+		if err != nil {
+			// We failed to send the request, the target must be dead.
+			log.DebugContext(ctx, "Failed to forward SSH request", "request_type", req.Type, "error", err)
+			// We must first send a reply if one was expected.
+			if req.WantReply {
+				req.Reply(false, nil)
+			}
+			// Close both connections or channels to clean up but we must
+			// continue handling requests on the chan until it is closed by
+			// crypto/ssh.
+			closeRequestSources()
+			continue
```

Comment:
`(*ssh.Request).Reply` already checks for `WantReply`, we can blindly `_ = req.Reply(false, nil)`.


```suggestion
			// Close both connections or channels to clean up but we must
			// continue handling requests on the chan until it is closed by
			// crypto/ssh.
			closeRequestSources()
			_ = req.Reply(false, nil)
			continue
```